crates/storage/api.rs

```diff
@@ -1,361 +1,142 @@
-    /// Doesn't check if the account is stored
-    /// Used for internal store operations
-    fn open_storage_trie(
+    /// Returns an iterator over all key-value pairs with the given prefix.
+    fn prefix_iterator(
         &self,
-        hashed_address: H256,
-        storage_root: H256,
-    ) -> Result<Trie, StoreError>;
-
-    /// Obtain a state trie from the given state root
-    /// Doesn't check if the state root is valid
-    /// Used for internal store operations
-    fn open_state_trie(&self, state_root: H256) -> Result<Trie, StoreError>;
-
-    /// Obtain a state trie locked for reads from the given state root
-    /// Doesn't check if the state root is valid
-    /// Used for internal store operations
-    fn open_locked_state_trie(&self, state_root: H256) -> Result<Trie, StoreError> {
-        self.open_state_trie(state_root)
-    }
+        table: &'static str,
+        prefix: &[u8],
+    ) -> Result<Box<dyn Iterator<Item = PrefixResult> + '_>, StoreError>;
+}
 
-    /// Obtain a read-locked storage trie from the given address and storage_root
-    /// Doesn't check if the account is stored
-    /// Used for internal store operations
-    fn open_locked_storage_trie(
-        &self,
-        hashed_address: H256,
-        storage_root: H256,
-    ) -> Result<Trie, StoreError> {
-        self.open_storage_trie(hashed_address, storage_root)
+/// Read-write transaction interface.
+///
+/// Extends [`StorageRoTx`] with methods to modify the database.
+/// Changes are not persisted until [`commit()`](StorageRwTx::commit) is called.
+pub trait StorageRwTx: StorageRoTx + Send {
+    /// Stores a key-value pair in the specified table.
+    fn put(&mut self, table: &'static str, key: &[u8], value: &[u8]) -> Result<(), StoreError> {
+        self.put_batch(vec![(table, key.to_vec(), value.to_vec())])
     }
 
-    async fn forkchoice_update(
-        &self,
-        new_canonical_blocks: Option<Vec<(BlockNumber, BlockHash)>>,
-        head_number: BlockNumber,
-        head_hash: BlockHash,
-        safe: Option<BlockNumber>,
-        finalized: Option<BlockNumber>,
-    ) -> Result<(), StoreError>;
-
-    fn get_receipts_for_block(&self, block_hash: &BlockHash) -> Result<Vec<Receipt>, StoreError>;
-
-    // Snap State methods
-
-    /// Sets the hash of the last header downloaded during a snap sync
-    async fn set_header_download_checkpoint(&self, block_hash: BlockHash)
+    /// Stores multiple key-value pairs in the specified table within the transaction.
+    fn put_batch(&mut self, batch: Vec<(&'static str, Vec<u8>, Vec<u8>)>)
     -> Result<(), StoreError>;
 
-    /// Gets the hash of the last header downloaded during a snap sync
-    async fn get_header_download_checkpoint(&self) -> Result<Option<BlockHash>, StoreError>;
+    /// Removes a key-value pair from the specified table.
+    fn delete(&mut self, table: &'static str, key: &[u8]) -> Result<(), StoreError>;
```

Comment:
I think it makes sense to have a `delete_batch` function too. Maybe even have `put` and `delete` be batch functions here, with _maybe_ a default `put_one`/`delete_one`.